README.txt
This package provides a FAKE mongoDB implementation and is quite handy for
testing.


CHANGES.txt
=======
CHANGES
=======


1.0.2 (unreleased)
------------------

- ...


1.0.1 (2015-03-17)
------------------

- bugfix: reflect changes in pymongo > 2.8.0. The dta basename checker method
  _check_name method was moved from pymongo.database to pymongo.helpers. Make
  import work for all pymongo versions.


1.0.0 (2013-03-06)
------------------

- Added support for Python 3.3.

- Dropped support for Python 2.4 and 2.5.


0.2.1 (2013-02-02)
------------------

- added realistic FakeCollection.remove return values

- bugfix around FakeCollection.save


0.2.0 (2012-12-16)
------------------

- feature: support pymongo 2.4 API and support new safe write concern

- include m01.stub for testing mongofake components and compare with real
  pymongo output. Use --all option for real pymongo and mongodb tests.

- implemented some basic tests and run them with m01.stub and m01.mongofake
  libraries

- implemented correct response for collection update method based on new
  pymongo 2.4. More will follow if more tests get implemented

- switch from lovely.importchecker to p01.recipe.setup importchecker which
  doesn't warn about bad svn url during buildout process


0.1.1 (2012-12-10)
------------------

- inproved dictify method, also convert single SON object to dict

- add some initial tests


0.1.0 (2012-05-22)
------------------

- First independent release from m01.mongo.testing


src/m01/mongofake/tests.py
##############################################################################
#
# Copyright (c) 2012 Zope Foundation and Contributors.
# All Rights Reserved.
#
# This software is subject to the provisions of the Zope Public License,
# Version 2.1 (ZPL).  A copy of the ZPL should accompany this distribution.
# THIS SOFTWARE IS PROVIDED "AS IS" AND ANY AND ALL EXPRESS OR IMPLIED
# WARRANTIES ARE DISCLAIMED, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED
# WARRANTIES OF TITLE, MERCHANTABILITY, AGAINST INFRINGEMENT, AND FITNESS
# FOR A PARTICULAR PURPOSE.
#
##############################################################################
"""
"""
import re
import unittest
import doctest

from zope.testing.renormalizing import RENormalizing

import m01.mongofake.testing

try:
    import m01.stub
except ImportError:
    has_m01_stub = False
else:
    has_m01_stub = True

PY_COMPAT = RENormalizing([
    # Python 3 unicode removed the "u".
    (re.compile("u('.*?')"),
     r"\1"),
    (re.compile('u(".*?")'),
     r"\1"),
    # Python 3 adds module.
    (re.compile('datetime.datetime'),
     r"datetime"),
   ])

CHECKER = RENormalizing([
   (re.compile('connectionId'), '...'),
   ])


FAKE_CHECKER = RENormalizing([
   (re.compile('FakeMongoClient'), 'MongoClient'),
   (re.compile('FakeDatabase'), 'Database'),
   (re.compile('FakeCollection'), 'Collection'),
   (re.compile("'connectionId': [0-9]+"), r"'connectionId': ..."),
   ])


def test_suite():
    """This test suite will run the tests with the fake and a real mongodb and
    make sure both output are the same.
    """
    suites = []
    append = suites.append

    # real mongo database tests using m01.stub using level 2 tests (--all)
    testNames = ['testing.txt',
                 ]
    if has_m01_stub:
        for name in testNames:
            suite = unittest.TestSuite((
                doctest.DocFileSuite(name,
                    setUp=m01.mongofake.testing.setUpStubMongo,
                    tearDown=m01.mongofake.testing.tearDownStubMongo,
                    optionflags=doctest.NORMALIZE_WHITESPACE|doctest.ELLIPSIS,
                    checker=PY_COMPAT+CHECKER),
                ))
            suite.level = 2
            append(suite)

    # fake mongo database tests using FakeMongoClient
    for name in testNames:
        append(
            doctest.DocFileSuite(name,
                setUp=m01.mongofake.testing.setUpFakeMongo,
                tearDown=m01.mongofake.testing.tearDownFakeMongo,
                optionflags=doctest.NORMALIZE_WHITESPACE|doctest.ELLIPSIS,
                checker=PY_COMPAT+FAKE_CHECKER),
        )

    # additional non mongodb tests
    append(
        doctest.DocFileSuite(
            'README.txt',
            checker=PY_COMPAT,
            optionflags=doctest.NORMALIZE_WHITESPACE|doctest.ELLIPSIS),
    )

    # return test suite
    return unittest.TestSuite(suites)


if __name__=='__main__':
    unittest.main(defaultTest='test_suite')


src/m01/mongofake/testing/__init__.py
##############################################################################
#
# Copyright (c) 2012 Zope Foundation and Contributors.
# All Rights Reserved.
#
# This software is subject to the provisions of the Zope Public License,
# Version 2.1 (ZPL).  A copy of the ZPL should accompany this distribution.
# THIS SOFTWARE IS PROVIDED "AS IS" AND ANY AND ALL EXPRESS OR IMPLIED
# WARRANTIES ARE DISCLAIMED, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED
# WARRANTIES OF TITLE, MERCHANTABILITY, AGAINST INFRINGEMENT, AND FITNESS
# FOR A PARTICULAR PURPOSE.
#
##############################################################################
"""
$Id$
"""
__docformat__ = "reStructuredText"

import os

import pymongo

import m01.stub.testing

import m01.mongofake


# mongo db name used for testing
TEST_DB_NAME = 'm01_mongofake_database'


###############################################################################
#
# test helper methods
#
###############################################################################

_testClient = None

def getTestClient():
    return _testClient


def getTestDatabase():
    client = getTestClient()
    return client[TEST_DB_NAME]


def getTestCollection(collectionName='test'):
    client = getTestClient()
    db = client[TEST_DB_NAME]
    return db[collectionName]


def dropTestDatabase():
    client = getTestClient()
    client.drop_database(TEST_DB_NAME)


###############################################################################
#
# test setup methods
#
###############################################################################

# fake mongodb setup
def setUpFakeMongo(test=None):
    """Setup fake (singleton) mongo client"""
    global _testClient
    host = 'localhost'
    port = 45017
    _testClient = m01.mongofake.fakeMongoClient(host, port)


def tearDownFakeMongo(test=None):
    """Tear down fake mongo client"""
    global _testClient
    _testClient = None


# stub mongodb server
def setUpStubMongo(test=None):
    """Setup real empty mongodb"""
    host = 'localhost'
    port = 45017
    sandBoxDir = os.path.join(os.path.dirname(__file__), 'sandbox')
    m01.stub.testing.startMongoDBServer(host, port, sandBoxDir=sandBoxDir)
    # ensure that we use a a real MongoClient
    global _testClient
    _testClient = pymongo.MongoClient(host, port)


def tearDownStubMongo(test=None):
    """Tear down real mongodb"""
    sleep = 0.5
    m01.stub.testing.stopMongoDBServer(sleep)


src/m01/mongofake/testing.py
##############################################################################
#
# Copyright (c) 2012 Zope Foundation and Contributors.
# All Rights Reserved.
#
# This software is subject to the provisions of the Zope Public License,
# Version 2.1 (ZPL).  A copy of the ZPL should accompany this distribution.
# THIS SOFTWARE IS PROVIDED "AS IS" AND ANY AND ALL EXPRESS OR IMPLIED
# WARRANTIES ARE DISCLAIMED, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED
# WARRANTIES OF TITLE, MERCHANTABILITY, AGAINST INFRINGEMENT, AND FITNESS
# FOR A PARTICULAR PURPOSE.
#
##############################################################################
"""Testing Support
"""
import os
import pymongo
import m01.mongofake

# mongo db name used for testing
TEST_DB_NAME = 'm01_mongofake_database'

###############################################################################
#
# test helper methods
#
###############################################################################

_testClient = None

def getTestClient():
    return _testClient


def getTestDatabase():
    client = getTestClient()
    return client[TEST_DB_NAME]


def getTestCollection(collectionName='test'):
    client = getTestClient()
    db = client[TEST_DB_NAME]
    return db[collectionName]


def dropTestDatabase():
    client = getTestClient()
    client.drop_database(TEST_DB_NAME)


###############################################################################
#
# test setup methods
#
###############################################################################

# fake mongodb setup
def setUpFakeMongo(test=None):
    """Setup fake (singleton) mongo client"""
    global _testClient
    host = 'localhost'
    port = 45017
    _testClient = m01.mongofake.fakeMongoClient(host, port)


def tearDownFakeMongo(test=None):
    """Tear down fake mongo client"""
    global _testClient
    _testClient = None


# stub mongodb server
def setUpStubMongo(test=None):
    """Setup real empty mongodb"""
    host = 'localhost'
    port = 45017
    sandBoxDir = os.path.join(os.path.dirname(__file__), 'sandbox')
    import m01.stub.testing
    m01.stub.testing.startMongoDBServer(host, port, sandBoxDir=sandBoxDir)
    # ensure that we use a a real MongoClient
    global _testClient
    _testClient = pymongo.MongoClient(host, port)


def tearDownStubMongo(test=None):
    """Tear down real mongodb"""
    sleep = 0.5
    import m01.stub.testing
    m01.stub.testing.stopMongoDBServer(sleep)
